### ISSUE TITLE
How to grant privilege per LDAP User

### ISSUE THREAD

**Issue by mattiZed** [dalibo/ldap2pg#444]
Hi!

First of, thanks for this awesome project.

I'm trying to achieve something on a development cluster what I think should be fairly simple:

Say I want each ldap-user that is in group "sql-user" to have full access to their own "experimental" schema, and read access to the rest. I started with the base ldap2pg.yml that you provide and added a grant section.

My problem is now, I get somewhat the "cross-product" of ldap users and experimental schemas.

The output for two users in ldap group "sql-users" would look something like this:

```
Would grant __create_on_schemas__ on my_db.experimental_alice to alice
Would grant __create_on_schemas__ on my_db.experimental_bob to alice
Would grant __create_on_schemas__ on my_db.experimental_alice to bob
Would grant __create_on_schemas__ on my_db.experimental_bob to bob
```

where I would expect to just have:

```
Would grant __create_on_schemas__ on my_db.experimental_alice to alice
Would grant __create_on_schemas__ on my_db.experimental_bob to bob
```

I don't quite understand what's happening, I hope you can correct my ldap2pg.yml accordingly.

Also, currently I have to create the "experimental_USER" schema manually, outside ldap2pg - is there a way to create schemas for each user in ldap2pg as well?

This is with ldap2pg==5.8.

Help much appreciated :)

## ldap2pg.yml

```
version: 5

postgres:
  databases_query: |
    SELECT datname FROM pg_database
    WHERE datallowconn IS TRUE
    AND datname = 'my_db';

  managed_roles_query: |
    SELECT DISTINCT role.rolname
    FROM pg_roles AS role
    LEFT OUTER JOIN pg_auth_members AS ms ON ms.member = role.oid
    LEFT OUTER JOIN pg_roles AS ldap_roles
      ON ldap_roles.rolname = 'ldap_roles' AND ldap_roles.oid = ms.roleid
    WHERE role.rolname IN ('ldap_roles')
        OR ldap_roles.oid IS NOT NULL
    ORDER BY 1;

  owners_query: |
    SELECT DISTINCT role.rolname
    FROM pg_catalog.pg_roles AS role
    JOIN pg_catalog.pg_auth_members AS ms ON ms.member = role.oid
    JOIN pg_catalog.pg_roles AS owners
      ON owners.rolname = 'owners' AND owners.oid = ms.roleid
    ORDER BY 1;

  schemas_query: |
    SELECT nspname FROM pg_catalog.pg_namespace
    WHERE nspname NOT LIKE 'pg_%' AND nspname <> 'information_schema'
    ORDER BY 1;


privileges:
  # Define ro privilege group with read-only grants
  ro:
  - __connect__
  - __select_on_tables__
  - __select_on_sequences__
  - __usage_on_schemas__
  - __usage_on_types__

  # rw privilege group lists write-only grants
  rw:
  - __temporary__
  - __all_on_tables__
  - __all_on_sequences__

  # ddl privilege group lists DDL only grants.
  ddl:
  - __create_on_schemas__


sync_map:
- description: "Setup static roles and grants."
  roles:
  - names:
    - ldap_roles
    - readers
    options: NOLOGIN
  - names: writers
    parent: readers
    options: NOLOGIN
  - name: owners
    parent: writers
    options: NOLOGIN

  grant:
  - privilege: ro
    role: readers
    schemas: __all__
  - privilege: rw
    role: writers
    schemas: __all__

- description: "Query LDAP to create users"
  ldapsearch:
    base: ou=not,dc=your,dc=concern
    filter: "(cn=sql-user)"
  roles:
    # Create role "username" & assign parent roles "ldap_roles" and "readers"
    - name: "{memberUid}"
      options: LOGIN
      parent:
      - ldap_roles
      - readers
      comment: "From LDAP group {dn}"
  grant:
    # Grant ddl privilege on schema "experimental_username"
    privilege: ddl
    role: "{memberUid}"
    schema: "experimental_{memberUid}"
  # grant:
  #   # Grant read-write privilege on schema "experimental_username"
  #   privilege: rw
  #   role: "{memberUid}"
  #   schema: "experimental_{memberUid}"
```

**Comment by bersace** [comment]
Hi @mattiZed , thanks for reaching.

What is the actual LDAP entry ? ldap2pg does a cartesian product of multiple values. If you have multiple `memberUid` you'll have such behaviour. It's described at https://ldap2pg.readthedocs.io/en/latest/ldap/

Can you confirm the output of cn=sql-user with ldapsearch ?

**Comment by mattiZed** [comment]
Hi @bersace , thank you for the quick reply.

Turns out I didn't understand much about how LDAP was working in the first place :)

I did some investigation and my query just returns a single entry with all the memberUids that make the ldap group. So for me to fix this problem, I need to create an LDAP query that returns a list of users which are part of the ldap group.

Unforntunately, our LDAP configuration does not currently support that due to some issue with the `memberOf` attribute, but this is a problem on our end.
